src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -167,6 +169,34 @@ export class WorkspaceSavedObjectsClientWrapper {
       return await wrapperOptions.client.find<T>(options);
     };
 
+    const addToWorkspacesWithPermissionControl = async (
+      objects: SavedObjectsShareObjects[],
+      targetWorkspaces: string[],
+      options: SavedObjectsAddToWorkspacesOptions = {}
+    ) => {
+      const objectToBulkGet = await wrapperOptions.client.bulkGet(objects, options);
+
+      // target workspaces
+      await this.validateMultiWorkspacesPermissions(
+        targetWorkspaces,
+        wrapperOptions.request,
+        WorkspacePermissionMode.Write
+      );
+
+      // saved_objects
+      const permitted = await this.permissionControl.validateSavedObjects(
+        objectToBulkGet.saved_objects,
+        WorkspacePermissionMode.Write,
+        wrapperOptions.request
+      );
+
+      if (!permitted) {
```

Comment:
Is that means to add one saved objects to workspace, we need two write permission, one for workspace write, another for saved objects write? Will the saved objects extends the workspace permission? If yes, we do we need verify it again?

Reply:
For target workspace, it will need write permission to add saved objects into it.

For saved objects want to add to target workspace. Here have two cases, 
1. have `library_write` permission of saved object workspaces.
2. have `write` permission to saved object itself, individual ACL.

I just consolidate them into one check, saved object must have write permission.

> one for workspace write

Do you mean this validation is on source workspace?

Reply:
right, we need to verify both source and target workspace.